api/src/main/resources/Java/libraries/okhttp-gson/api.mustache

```diff
@@ -0,0 +1,277 @@
+import {{invokerPackage}}.ApiResponse;
+import {{invokerPackage}}.Configuration;
+import {{invokerPackage}}.Pair;
+import {{invokerPackage}}.ProgressRequestBody;
+import {{invokerPackage}}.ProgressResponseBody;
+{{#performBeanValidation}}
+import {{invokerPackage}}.BeanValidationException;
+{{/performBeanValidation}}
+
+import com.google.gson.reflect.TypeToken;
+
+import java.io.IOException;
+
+{{#useBeanValidation}}
+import javax.validation.constraints.*;
+{{/useBeanValidation}}
+{{#performBeanValidation}}
+import javax.validation.ConstraintViolation;
+import javax.validation.Validation;
+import javax.validation.ValidatorFactory;
+import javax.validation.executable.ExecutableValidator;
+import java.util.Set;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+{{/performBeanValidation}}
+
+{{#imports}}import {{import}};
+{{/imports}}
+
+{{! Begin AoU-specific additions }}
+import io.opencensus.common.Scope;
+import io.opencensus.trace.AttributeValue;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Status;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import java.lang.reflect.Type;
+{{! End AoU-specific additions }}
+{{^fullJavaUtil}}
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+{{/fullJavaUtil}}
+
+{{#operations}}
+public class {{classname}} {
+    private static final Tracer tracer = Tracing.getTracer();
+    private ApiClient {{localVariablePrefix}}apiClient;
+
+    public {{classname}}() {
+        this(Configuration.getDefaultApiClient());
+    }
+
+    public {{classname}}(ApiClient apiClient) {
+        this.{{localVariablePrefix}}apiClient = apiClient;
+    }
+
+    public ApiClient getApiClient() {
+        return {{localVariablePrefix}}apiClient;
+    }
+
+    public void setApiClient(ApiClient apiClient) {
+        this.{{localVariablePrefix}}apiClient = apiClient;
+    }
+
+    {{#operation}}
+    /**
+     * Build call for {{operationId}}{{#allParams}}
+     * @param {{paramName}} {{description}}{{#required}} (required){{/required}}{{^required}} (optional{{#defaultValue}}, default to {{{.}}}{{/defaultValue}}){{/required}}{{/allParams}}
+     * @param progressListener Progress listener
+     * @param progressRequestListener Progress request listener
+     * @return Call to execute
+     * @throws ApiException If fail to serialize the request body object
+     */
+    public com.squareup.okhttp.Call {{operationId}}Call({{#allParams}}{{{dataType}}} {{paramName}}, {{/allParams}}final ProgressResponseBody.ProgressListener progressListener, final ProgressRequestBody.ProgressRequestListener progressRequestListener) throws ApiException {
+        Object {{localVariablePrefix}}localVarPostBody = {{#bodyParam}}{{paramName}}{{/bodyParam}}{{^bodyParam}}null{{/bodyParam}};
+
+        // create path and map variables
+        String {{localVariablePrefix}}localVarPath = "{{{path}}}"{{#pathParams}}
```

Comment:
Do all these string acrobatics need to be generated, or could you write a utility function that gets called from here?

Reply:
/noideadog.

I'll resolve most of this and the following comment threads, as I don't want to dig into too many details around the swagger-codegen project's code style or design decisions. I've tried to limit the scope of the template changes made in this PR – here's the old and new template contents, with a diff: https://gist.github.com/gjuggler/0a485915da5e68e26150e9decf4a7d67

As I dug into this, I did realize / remember that the actual version of api.mustache that I'd copied over is the version from the 2.2.3 release, which is the version we import in Gradle and was released in ~July 2017. Here is the file link: https://github.com/swagger-api/swagger-codegen/blob/v2.2.3/modules/swagger-codegen/src/main/resources/Java/libraries/okhttp-gson/api.mustache

I've added some more comments to the top of the file, as well as to where we import swagger-codegen, about this gnarly implicit dependency. Unfortunately I don't think there's a clean & nice way to make it easier to override a specific part of this codegen template. If we upgrade our version of swagger-codegen, it's almost certainly going to require us to update our template overrides to be compatible with the newer version. An interesting question is whether this add'l friction to upgrade would be worth the benefit of having our ability to modify how Swagger is generating things.